x-pack/plugins/fleet/server/services/agent_policies/package_policies_to_agent_permissions.ts

```diff
@@ -49,112 +50,114 @@ export async function storedPackagePoliciesToAgentPermissions(
-          // - Osquery manager doesn't store the `data_stream` metadata in
-          //   `packagePolicy.inputs`, so we will use _all_ data_streams from
-          //   the package.
-          dataStreamsForPermissions = dataStreams;
-          break;
-
-        default:
-          // - Normal packages store some of the `data_stream` metadata in
-          //   `packagePolicy.inputs[].streams[].data_stream`
-          // - The rest of the metadata needs to be fetched from the
-          //   `data_stream` object in the package. The link is
-          //   `packagePolicy.inputs[].type == dataStreams.streams[].input`
-          // - Some packages (custom logs) have a compiled dataset, stored in
-          //   `input.streams.compiled_stream.data_stream.dataset`
-          dataStreamsForPermissions = packagePolicy.inputs
-            .filter((i) => i.enabled)
-            .flatMap((input) => {
-              if (!input.streams) {
-                return [];
-              }
-
-              const dataStreams_: DataStreamMeta[] = [];
-
-              input.streams
-                .filter((s) => s.enabled)
-                .forEach((stream) => {
-                  if (!('data_stream' in stream)) {
-                    return;
-                  }
-
-                  const ds: DataStreamMeta = {
-                    type: stream.data_stream.type,
-                    dataset:
-                      stream.compiled_stream?.data_stream?.dataset ?? stream.data_stream.dataset,
-                  };
-
-                  if (stream.data_stream.elasticsearch) {
-                    ds.elasticsearch = stream.data_stream.elasticsearch;
-                  }
-
-                  dataStreams_.push(ds);
-                });
-
-              return dataStreams_;
-            });
-      }
-
-      let clusterRoleDescriptor = {};
-      const cluster = packagePolicy?.elasticsearch?.privileges?.cluster ?? [];
-      if (cluster.length > 0) {
-        clusterRoleDescriptor = {
-          cluster,
-        };
-      }
-
-      return [
-        packagePolicy.id,
-        {
-          indices: dataStreamsForPermissions.map((ds) =>
-            getDataStreamPrivileges(ds, packagePolicy.namespace)
-          ),
-          ...clusterRoleDescriptor,
-        },
-      ];
+  const permissionEntries = (packagePolicies as PackagePolicy[]).map(async (packagePolicy) => {
+    if (!packagePolicy.package) {
+      throw new Error(`No package for package policy ${packagePolicy.name ?? packagePolicy.id}`);
+    }
+
+    const pkg = packageInfoCache.get(pkgToPkgKey(packagePolicy.package))!;
+
+    // Special handling for Universal Profiling packages, as it does not use data streams _only_,
+    // but also indices that do not adhere to the convention.
+    if (
+      pkg.name === FLEET_UNIVERSAL_PROFILING_SYMBOLIZER_PACKAGE ||
+      pkg.name === FLEET_UNIVERSAL_PROFILING_COLLECTOR_PACKAGE
+    ) {
+      return universalProfilingPermissions(packagePolicy.id);
```

Comment:
Does this work correctly if the `storedPackagePoliciesToAgentPermissions` return type is a promise?

Reply:
Yes, i cleaned it up because it doesn't need to be a promise. Anything left to do here?

Reply:
@juliaElastic I pushed one more commit where I remove the remaining async behaviour (and cleaned up some of the types).